src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java

```diff
@@ -453,32 +455,73 @@ private CompletableFuture<Response> checkFor407(ExchangeImpl<T> ex, Throwable t,
+        }
+
         return ex.getResponseAsync(parentExecutor)
+                .orTimeout(responseTimeout, TimeUnit.SECONDS)
+                .handle((response, exception) -> {
+                    CompletableFuture<Response> cf;
+                    if (exception != null) {
+                        if (exception instanceof TimeoutException) {
+                            // Timeout has occurred, this SHOULD result in the request being sent anyway
+                            // return the null response to be handled below
+                            if (debug.on())
+                                debug.log("Request timeout waiting for response");
+                            cf = CompletableFuture.completedFuture(response);
+                        } else {
+                            // All other exceptions get returned as failedFuture
+                            cf = CompletableFuture.failedFuture(exception);
+                        }
+                        return cf;
+                    }
+                    // If there was no exception then return the response
+                    cf = CompletableFuture.completedFuture(response);
+                    return cf;
+                }).thenCompose(Function.identity())
                 .thenCompose((Response r1) -> {
-            Log.logResponse(r1::toString);
-            int rcode = r1.statusCode();
-            if (rcode == 100) {
-                Log.logTrace("Received 100-Continue: sending body");
-                if (debug.on()) debug.log("Received 100-Continue for %s", r1);
-                CompletableFuture<Response> cf =
-                        exchImpl.sendBodyAsync()
-                                .thenCompose(exIm -> exIm.getResponseAsync(parentExecutor));
-                cf = wrapForUpgrade(cf);
-                cf = wrapForLog(cf);
-                return cf;
-            } else {
-                Log.logTrace("Expectation failed: Received {0}",
-                        rcode);
-                if (debug.on()) debug.log("Expect-Continue failed (%d) for: %s", rcode, r1);
-                if (upgrading && rcode == 101) {
-                    IOException failed = new IOException(
-                            "Unable to handle 101 while waiting for 100");
-                    return MinimalFuture.failedFuture(failed);
-                }
-                exchImpl.expectContinueFailed(rcode);
-                return MinimalFuture.completedFuture(r1);
-            }
-        });
+                    // The response will only be null if there was a timeout
+                    // send body regardless
+                    if (r1 == null) {
+                        if (debug.on())
+                            debug.log("Setting ExpectTimeoutRaised and sending request body");
+                        exchImpl.setExpectTimeoutRaised(true);
+                        CompletableFuture<Response> cf =
+                                exchImpl.sendBodyAsync()
```

Comment:
When a timeout occurs due to `orTimeout()` step of a CompletableFuture pipeline, the subsequent stages in that pipeline get invoked on a thread that's internal to the `java.util.concurrent` code and isn't the one that's part of the executor that's configured on our HttpClient.
Looking at `sendBodyAsync()` which gets invoked if there's a timeout, I suspect it's OK if that method gets invoked through this internal thread. Daniel, do you see any issues if this internal thread ends up calling the `sendBodyAsync()` here?

Reply:
There should be no issue. The future returned to the caller of sendAsync has the appropriate calls to make sure user dependent actions are run in the FJP.